### ISSUE TITLE
Mapping and using SQL Server stored procedures with EF Core Power Tools

### ISSUE THREAD

**Issue by ErikEJ** [ErikEJ/erikej.github.io#28]

**Comment by dfjackson** [comment]
How can I enable this feature?  I'm using the latest version 2.5.513.0 but there's no option to "Generate map stored procedure mappings".  Here is a screenshot of the reverse engineering advanced options which doesn't have this option:
![image](https://user-images.githubusercontent.com/3920914/108561249-e58bbd00-72cb-11eb-9f25-ac08e7b0d42f.png)

**Comment by ErikEJ** [comment]
@dfjackson You simply just pick them, and use them like [described in the docs here](https://github.com/ErikEJ/EFCorePowerTools/wiki/Reverse-Engineering#sql-server-stored-procedures).

**Comment by dfjackson** [comment]
Ok thanks.  I noticed that in the Choose Database Objects dialog, the OK button is greyed out when you select stored procedures unless you also select some objects other than stored procedures so maybe this is a bug.

**Comment by ErikEJ** [comment]
@dfjackson It is actually by design, as you need a DbContext to connect to the stored procedures class, as the EF Core tooling does not allow a DbContext without any tables/views.

**Comment by markodikic** [comment]
I have stored procedure that has 3 select statements. First two select statements put data into temp tables and the third returns actual result (results is multicolumn data). Where I use EF Core Power Tools to generate stored procedure, I get following message:

> Unable to get result set shape for <PROCEDURE NAME>
> The metadata could not be determined because statement <STATEMENT PART> uses a temp table.

Is there a way to instruct the tool to use only final SELECT statement to get result shape or to rewrite procedure somehow but to still use temp tables?

**Comment by ErikEJ** [comment]
@markodikic No, but an empty XxxResult.cs class should have been implemented, you can add a XxxResult.Partial.cs partial class, and add properties for the known columns in that, then you should be able to get the result set.

**Comment by markodikic** [comment]
Thanks, Erik.

**Comment by ErikEJ** [comment]
I will get that added to the docs!

**Comment by mahldcat** [comment]
This functionality is _almost_ what I was looking for---My main concern was to find a tool that would allow me to reverse engineer not only the DB tables/views but the sprocs as well.

This does exactly that, but I'm now interested in replicating the reverse engineering steps on the command line?

Ultimate goal is to make this part of a CI/CD build process that would allow me to model the DB, and turn this code (along with any "repair" code) into its own assembly?

I was looking at the source code, and it looks like there was a command line to extract the table/sprocs/views etc into some json, and was looking at the main EFCorePowerTools code in the Handlers\ReverseEngineer section of the code but got bogged down pretty quickly?

**Comment by ErikEJ** [comment]
@mahldcat Yes, there is a undocumented/unsupported command line interface, generally I would advise against shipping generated code without having it reviewed by a human!

Having said that, I have a general description of the tool here, that you may find useful:
https://github.com/ErikEJ/EFCorePowerTools/issues/340#issuecomment-670929754

**Comment by nw5428** [comment]
I would have rather of added the sproc result classes to the dbcontext and also use parameter interpolation instead of using sqlParameter.

**Comment by ErikEJ** [comment]
@airwedge1 
>> added the sproc result classes to the dbcontext

I fail to se the benefit of that, you cannot compose over stored procedure results like you can with tables and views.

>>  parameter interpolation instead of using sqlParameter.

That will increase the built-in risk of EF Core picking the wrong parameter properties (length, DbType)

**Comment by dchillman123** [comment]
after I use the tool for stored procedures that return a result set, I manually update the generated code with specific database objects, but i have to remember to do that every time I run the tool.

eg. change
public virtual async Task<List<uspGetBasicScreenQuestionsResult>> uspGetBasicScreenQuestionsAsync
to
public virtual async Task<List<BasicScreenObject>> uspGetBasicScreenQuestionsAsync

am i missing something in the tool that lets me do that just once?

**Comment by ErikEJ** [comment]
@dchillman123 I do not understand your question, before and after look identical.

**Comment by dchillman123** [comment]
ugg - guess i need to actually read before i push the button.  I meant to show that I am manually changing the object after the code is generated (uspGetScreenAssessmentDetailsResult is replaced by ScreenAssessmentDetailsView)

generated code
`public virtual async Task<List<uspGetScreenAssessmentDetailsResult>> uspGetScreenAssessmentDetailsAsync(long? ScreenAssessmentId, OutputParameter<int> returnValue = null, CancellationToken cancellationToken = default)
        {
            var parameterreturnValue = new SqlParameter
            {
                ParameterName = "returnValue",
                Direction = System.Data.ParameterDirection.Output,
                SqlDbType = System.Data.SqlDbType.Int,
            };

            var sqlParameters = new []
            {
                new SqlParameter
                {
                    ParameterName = "ScreenAssessmentId",
                    Value = ScreenAssessmentId ?? Convert.DBNull,
                    SqlDbType = System.Data.SqlDbType.BigInt,
                },
                parameterreturnValue,
            };
            var _ = await _context.SqlQueryAsync<uspGetScreenAssessmentDetailsResult>("EXEC @returnValue = [dbo].[uspGetScreenAssessmentDetails] @ScreenAssessmentId", sqlParameters, cancellationToken);

            returnValue?.SetValue(parameterreturnValue.Value);

            return _;
        }`

how i update manually after generation
`public virtual async Task<List<ScreenAssessmentDetailsView>> uspGetScreenAssessmentDetailsAsync(long? ScreenAssessmentId, OutputParameter<int> returnValue = null, CancellationToken cancellationToken = default)
        {
            var parameterreturnValue = new SqlParameter
            {
                ParameterName = "returnValue",
                Direction = System.Data.ParameterDirection.Output,
                SqlDbType = System.Data.SqlDbType.Int,
            };

            var sqlParameters = new []
            {
                new SqlParameter
                {
                    ParameterName = "ScreenAssessmentId",
                    Value = ScreenAssessmentId ?? Convert.DBNull,
                    SqlDbType = System.Data.SqlDbType.BigInt,
                },
                parameterreturnValue,
            };
            var _ = await _context.SqlQueryAsync<ScreenAssessmentDetailsView>("EXEC @returnValue = [dbo].[uspGetScreenAssessmentDetails] @ScreenAssessmentId", sqlParameters, cancellationToken);

            returnValue?.SetValue(parameterreturnValue.Value);

            return _;
        }`

**Comment by ErikEJ** [comment]
Please mark all you text as code, still - cannot see any difference.

**Comment by ErikEJ** [comment]
@dchillman123 and why is renaming the result class important (if that is what you do) ?

**Comment by dchillman123** [comment]
yeah, i am just renaming the class to match an class in my solution. then i can directly define it in code

List<ScreenAssessmentDetailsView> lstMyDetails = await procedures.uspGetScreenAssessmentDetailsAsync(screenAssessmentId);

If I don't do that, I have to convert the list (with the generic class generated by the tool)  to a list with my  object class, which I can then pass to my view model.  

I guess it is more of a convenience issue

**Comment by ErikEJ** [comment]
I have trouble understanding your issue.

Have you tried the renaming feature?

(Or just rename your viewmodel class?)

**Comment by dchillman123** [comment]
well, that should work - i always forget to use the simplest solution.

**Comment by SandroRiz** [comment]
I have the same question of above...
The reverse engineer of a SP create always a new entity called like the sp. What about, if the SP returns a table already scaffolded and so you could reuse a just created model ? I remember for example Linq2SQL dbml designer where you could choose what type of entity (or scalar) the SP returns

To avoid renaming the class in the generated code, or worse, to automap to identical object, it will be useful...

thx

**Comment by ErikEJ** [comment]
@SandroRiz please create a proposal with more details in the Github repo

**Comment by stevedrew41075** [comment]
Hello, accessing tables has worked perfectly but I cannot seem to find a solution to running a stored procedure.  I used a very simple example from the wiki which you will see below but it blows up when running the stored procedure.   Thank you in advance for your assistance.

**Program.cs – dependency injection**
builder.Services.AddDbContext<MyContext>(options => options.UseSqlServer(builder.Configuration.GetConnectionString("DefaultConnection")))
builder.Services.AddScoped<IUnitOfWork, UnitOfWork>();


    public class UploadFactory : IUploadFactory
    {
        private IUnitOfWork UnitOfWork;
        private IUserFactory UserFactory;
        public UploadFactory(
            IUnitOfWork unitOfWork,
            IUserFactory userFactory)
        {
            this.UnitOfWork = unitOfWork;
            this.UserFactory = userFactory;
        }

**<skip some stuff … call this method>**
        private async Task<int> RunProc(int xmlID)
        {
            using (var db = new MyContext())
            {
                **await db.GetProcedures().up_SC_Load_XML_From_DatabaseAsync(xmlID);**
            }
            return 0;
        }


Created by EF Power Tools – it hits the bold line using the correct parameter value
        public virtual async Task<int> up_SC_Load_XML_From_DatabaseAsync(int? XMLRecNum, OutputParameter<int> returnValue = null, CancellationToken cancellationToken = default)
        {
            var parameterreturnValue = new SqlParameter
            {
                ParameterName = "returnValue",
                Direction = System.Data.ParameterDirection.Output,
                SqlDbType = System.Data.SqlDbType.Int,
            };

            var sqlParameters = new []
            {
                new SqlParameter
                {
                    ParameterName = "XMLRecNum",
                    Value = XMLRecNum ?? Convert.DBNull,
                    SqlDbType = System.Data.SqlDbType.Int,
                },
                parameterreturnValue,
            };
            **var _ = await _context.Database.ExecuteSqlRawAsync("EXEC @returnValue = [dbo].[up_SC_Load_XML_From_Database] @XMLRecNum", sqlParameters, cancellationToken);**

            returnValue?.SetValue(parameterreturnValue.Value);

            return _;
        }

The stored proc signature – it does not return a result set or return code
ALTER PROCEDURE [dbo].[up_SC_Load_XML_From_Database]
	@XMLRecNum			INTEGER

AS
…. Just does a bunch of exec statements inside

The error message:
System.InvalidOperationException
  HResult=0x80131509
  Message=No database provider has been configured for this DbContext. A provider can be configured by overriding the 'DbContext.OnConfiguring' method or by using 'AddDbContext' on the application service provider. If 'AddDbContext' is used, then also ensure that your DbContext type accepts a DbContextOptions<TContext> object in its constructor and passes it to the base constructor for DbContext.
  Source=Microsoft.EntityFrameworkCore

UnitOfWork exposes just a few functions but depends on MyContext
    public interface IUnitOfWork : IDisposable
    {
        MyContext Context { get; }

        void Dispose(bool disposing);

        bool InTransaction { get; }

        ITransaction BeginTransaction();
    }

**Comment by ErikEJ** [comment]
The error message seems quite clear. If you want help, please provide a full  runnable repro project and share in the Power Tools repository.
@stevedrew41075

**Comment by stevedrew41075** [comment]
I'm a novice when it comes to DB Contexts and it would be difficult to provide a runnable version since it connects to a DB server not accessible via the internet.  Could you please give me a hint as to the possible solution or point me to a project that shows it in action?   Thanks for you help!

**Comment by ErikEJ** [comment]
It looks like you are doing your Dependency injection wrong, but without code it is impossible for me to see! 
@stevedrew41075

**Comment by stevedrew41075** [comment]
Hey Eric,  pointing me to a problem with the injection helped me solve it!  Thanks

**Comment by ravalds2016** [comment]
Hello Eric, in my project we already have a DBContext that has bunch of DBSets for tables in the database. I used this tool to access stored procedures from the same database. Your tool created a new context for the same database just for Stored Procedure access. Not sure if two DBContexts accessing the same DB is good or not. I do not see : efpt.config.json for original DBContext that pre-existed which I could use to add SPROCs to existing DBContext. Using EF core power tool reverse engineering for accessing stored procs from the same database creates another Db Context. Is this the only option available ? Any option to use existing DBContext to access stored procedures using EF core power tools without the config.json file ? I

**Comment by ErikEJ** [comment]
The tool no longer creates a separate DbContext. Not sure if that solvrs your issue. If not, please create an issue in the EF Core Power Tools repo.

@ravalds2016

**Comment by kjjpfeiffer** [comment]
Just a quick question on reverse engr of stored procs.  I have the IAcuityStoreContextProcedures file with definitions, but the OnModelCreating generates the lines
            OnModelCreatingGeneratedProcedures(modelBuilder);
            OnModelCreatingPartial(modelBuilder);

VS does not see an implementation of the first (nor do I).  I generated the context in the project root and the entities in .\Models.

**Comment by ErikEJ** [comment]
Please create an issue in the Power Tools  repo with proper reproduction steps or even better a repro project.

**Comment by kjjpfeiffer** [comment]
Turns out it was a one-time thing.  No errors were reported. I created a new project and only brought in one table (instead of the >100 tables and views) and all the sprocs needed.  This time both I...ContextProcedures.cs and ...ContextProcedures.cs were created.  Copied these to the larger project.

**Comment by ErikEJ** [comment]
@kjjpfeiffer OK ?!

**Comment by kjjpfeiffer** [comment]
My thought too.

**Comment by Amin-ir** [comment]
Hi, i've successfully added my database which contains tables and stored procedures to my asp.net core project through Power Tools with EF Core 5; but when i call the procedure and store its output to a viewbag, what i see in the View instead is this:
>System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Collections.Generic.List`1[WebApplication2.Models.DB.retTableResult],WebApplication2.Models.DB.schoolContextProcedures+<retTableAsync>d__3]
The store procedure code i made at db:
```
CREATE PROCEDURE retTable (@P INT)
AS
BEGIN
SELECT * FROM STU WHERE ID=@P;
END;
```
and this is what i wrote at the controller:
```
        public IActionResult Index()
        {
            schoolContext s = new schoolContext();
            schoolContextProcedures sp = new schoolContextProcedures(s);
            var t = sp.retTableAsync(20); // The built is perfectly done, but the value store instead of result set is weird
            ViewBag.test = t;      
            return View();
        }
```
How do i access the result set of the stored procedure?

**Comment by ErikEJ** [comment]
@Amin-ir you must use await when calling  the async method.

**Comment by matajack** [comment]
> I have the same question of above... The reverse engineer of a SP create always a new entity called like the sp. What about, if the SP returns a table already scaffolded and so you could reuse a just created model ? I remember for example Linq2SQL dbml designer where you could choose what type of entity (or scalar) the SP returns
> 
> To avoid renaming the class in the generated code, or worse, to automap to identical object, it will be useful...
> 
> thx

@ErikEJ is there a solution to the question asked by @SandroRiz regarding using an existing EF model for a generated SP in power tools?

**Comment by ErikEJ** [comment]
@markodikic @SandroRiz This has already been added a while ago: https://github.com/ErikEJ/EFCorePowerTools/wiki/Reverse-Engineering#sql-server-stored-procedures

**Comment by nw5428** [comment]
MataJack,

Check out https://github.com/ErikEJ/EFCorePowerTools/wiki/Reverse-Engineering#sql-server-stored-procedures


[created with MySignature.io]<https://kubicacorp.com/>
Nicholas Walker
Software Leader | Kubica
mobile:  248.924.1507
website: Kubicacorp.com<https://kubicacorp.com/>
email:    ***@***.******@***.***>

**Comment by pwolfson** [comment]
@ErikEJ It seems like the tool is able to load entities that are mapped to the stored procedure, but only the parent entity gets loaded. None of the related entities, such as child collections, are loaded. 

Is there currently a way in EF Core to execute a stored procedure and have it return a mapped entity, that is fully loaded with all its children, and tracked by Entity Framework?

**Comment by ErikEJ** [comment]
@pwolfson how would the SQL for that look?

**Comment by pwolfson** [comment]
@ErikEJ similar to EF Core 7 stored procedure mapping features such as UpdateUsingStoredProcedure. You create these maps for each Entity-To-Table relationship so that each Entity knows which stored procedure to run to materialize itself.

So with the classic Company and Employee example...for the Company we would setup a map using SelectUsingStoredProcedure and that stored procedure would be nothing more than SELECT * FROM Companies WHERE Id = @CompanyId. Next, since the Company has a Users collection, it would then find the SelectUsingStoredProcedure map for that Entity which would be nothing more Thant SELECT * FROM Users WHERE CompanyId = @CompanyId...this would be the foreign key.

**Comment by ErikEJ** [comment]
@pwolfson ok. Sounds complex. And certainly not possible at the moment.

**Comment by CaiusJard** [comment]
@pwolfson if you're getting back an expected entity with relation etc, then perhaps you can load all the related entities via Explicit Loading - https://learn.microsoft.com/en-us/ef/core/querying/related-data/explicit?source=recommendations - a job for your repo layer, or make an extension method for your context?

**Comment by andrenevesgomes** [comment]
Hi Erik,

First off, thanks for all your hard work on EF Core Power Tools – it's been an invaluable tool for my projects!

I have a quick question regarding reverse engineering behavior. When I reverse engineer tables or views from my SQL database, a single `DbContext` is generated as expected. However, when I reverse engineer a stored procedure, it seems to generate a separate `DbContext`. 

I'm curious if there's a reason for this distinction? Is it intentional for stored procedures to have a different `DbContext`, or am I missing something in my setup? I’ve checked the documentation but haven’t found much about this scenario.

Thanks again for your great tool, and I appreciate any insight you can provide!

**Comment by ErikEJ** [comment]
@andrenevesgomes it does not generate a seperate DbContext. Feel free to follow up in the EF Core Power Tools repo..

**Comment by HansBM** [comment]
@ErikEJ I think @andrenevesgomes means that:  When reverse engeneering tables, views, SPs and functions I get 3 partial DbContexts:

One for tables and views, for example FPEntities.cs in my case.
Then one for functions, named FPEntities.Functions.cs
And finally one for stored procedures named FPEntitiesProcedures.cs

It is no problem, I just think people are surprised by this.

**Comment by ErikEJ** [comment]
@HansBM Sure, this is by design.
